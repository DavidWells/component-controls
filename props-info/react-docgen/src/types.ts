import { ParserOptions } from '@babel/parser';

type ASTNode = object;

export interface Scope {
  lookup: (name: string) => Scope | undefined;
  lookupType: (name: string) => Scope | undefined;
  getBindings: () => { [key: string]: NodePath[] };
  getTypes: () => { [key: string]: NodePath[] };
  node: NodePath;
}

export interface NodePath {
  value: ASTNode | ASTNode[];
  node: ASTNode;
  parent: NodePath;
  parentPath: NodePath;
  scope: Scope;
  get(...x: (string | number)[]): NodePath;
  each(f: (p: NodePath) => any): any;
  map<T>(f: (p: NodePath) => T): T[];
  filter(f: (p: NodePath) => boolean): NodePath[];
  push(node: ASTNode): void;
}
type Documentation = any;

export type ResolverType = (
  ast: ASTNode,
  parser: { parse: (srource: string) => ASTNode },
) => NodePath | NodePath[];

export interface OptionsType {
  filename?: string;
  cwd?: string;
  babelrc?: boolean;
  babelrcRoots?: boolean;
  root?: boolean;
  rootMode?: boolean;
  configFile?: boolean;
  envName?: boolean;
  parserOptions?: ParserOptions;
}
export type HandlerType = (
  documentation: Documentation,
  definition: NodePath,
  parser: { parse: (source: string) => ASTNode },
) => void;

export interface RectDocgenOptions {
  resolver?: ResolverType;
  handlers?: HandlerType[];
  options?: OptionsType;
}
