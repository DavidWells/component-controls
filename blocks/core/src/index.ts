export * from './ActionBar';
export * from './BlockContainer';
export * from './Source';
export * from './Tabs';
export * from './ControlsTable';
