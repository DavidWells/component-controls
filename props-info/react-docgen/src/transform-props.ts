import {
  PropTypes,
  PropType,
  TypeInformatiom,
  TypeValue,
} from '@component-controls/specification';

const rdPropToCCProp = (rdProp: any): PropType => {
  const prop: Partial<PropType> = {};
  if (rdProp.description) {
    prop.description = rdProp.description;
  }
  if (rdProp.defaultValue !== null && rdProp.defaultValue !== undefined) {
    prop.defaultValue = rdProp.defaultValue.value || rdProp.defaultValue;
  }
  if (rdProp.required) {
    prop.required = rdProp.required;
  }
  let type: Partial<TypeInformatiom> = {};
  if (rdProp.type) {
    type = propTypeToCCType(rdProp.type);
  } else if (rdProp.tsType) {
    type = tsTypeToCCType(rdProp.tsType);
  } else if (rdProp.flowType) {
    type = tsTypeToCCType(rdProp.flowType);
  }
  prop.type = type as TypeInformatiom;
  return prop as PropType;
};
const propTypeToCCType = (dgType: any): TypeValue => {
  let type: TypeValue = {};
  if (dgType) {
    const typeName = dgType.name;
    if (dgType.required) {
      type.required = true;
    }
    switch (typeName) {
      case 'arrayOf':
      case 'array':
        type = {
          name: 'array',
        };
        if (dgType.value) {
          type.value = Object.keys(dgType.value).map(name => ({
            name: dgType.value[name],
          }));
        }
        break;
      case 'bool':
        type = {
          name: 'boolean',
        };
        break;
      case 'enum':
        type = {
          name: 'enum',
          value: dgType.value.map(({ value }: any) => {
            return {
              name: typeof value,
              value,
            };
          }),
        };
        break;
      case 'element':
      case 'elementType':
      case 'node':
        type = {
          name: 'any',
          raw: typeName,
        };
        break;

      case 'instanceOf':
        type = {
          name: 'object',
          value: dgType.value,
          raw: typeName,
        };
        break;
      case 'shape':
      case 'exact':
        type = {
          name: 'object',
          value: Object.keys(dgType.value).map(name => {
            const t: TypeValue = {
              name: dgType.value[name].name,
              value: name,
            };
            if (dgType.value[name].required) {
              t.required = true;
            }
            return t;
          }),
        };

        break;
      case 'union':
        type = {
          name: 'union',
          value: dgType.value.map((v: any) => propTypeToCCType(v)),
        };
        break;
      case 'func':
        type = {
          ...dgType,
          name: 'function',
        };
        break;

      default:
        type = dgType;
    }
  }
  return type;
};

const tsTypeToCCType = (dgType: any): TypeValue => {
  let type: TypeValue = {};
  if (dgType) {
    const typeName = dgType.name;
    if (dgType.required) {
      type.required = true;
    }
    switch (typeName) {
      case 'Array':
        type = {
          name: 'array',
          value: dgType.elements.map((element: any) => tsTypeToCCType(element)),
        };
        break;
      case 'signature':
        if (dgType.signature.arguments) {
          type = {
            name: 'function',
            value: tsTypeToCCType(dgType.signature.return),
            arguments: dgType.signature.arguments.map((v: any) =>
              tsTypeToCCType(v),
            ),
          };
        } else {
          type = {
            name: 'object',
            value: dgType.signature.properties.map((v: any) =>
              tsTypeToCCType(v),
            ),
          };
        }
        break;

      default:
        type = dgType;
    }
  }
  return type;
};
export const transformProps = (props: any): PropTypes | undefined => {
  return props
    ? Object.keys(props).reduce((acc, name) => {
        const rdProp: any = props[name];
        const prop = rdPropToCCProp(rdProp);
        return { ...acc, [name]: prop };
      }, {})
    : undefined;
};
