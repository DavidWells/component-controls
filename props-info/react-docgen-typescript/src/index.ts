import { ComponentInfo } from '@component-controls/specification';
import {
  extractDocgenTypescriptInfo,
  RectDocgenTypescriptOptions,
} from './react-docgen-typescript';
import { transformProps } from './transform-props';

export default (options?: RectDocgenTypescriptOptions) => {
  return async (
    fileName: string,
    componentName?: string,
  ): Promise<ComponentInfo | undefined> => {
    const propTable = extractDocgenTypescriptInfo(
      fileName,
      componentName,
      options,
    );
    return new Promise(resolve =>
      resolve(
        propTable
          ? {
              ...propTable,
              props: transformProps(propTable.props),
            }
          : undefined,
      ),
    );
  };
};
