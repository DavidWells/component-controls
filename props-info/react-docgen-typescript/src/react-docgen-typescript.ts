import { ParserOptions, withDefaultConfig } from 'react-docgen-typescript';
import { computeComponentName } from './defaults';
import { RectDocgenTypescriptOptions } from './types';

export const extractDocgenTypescriptInfo = (
  fileName: string,
  componentName?: string,
  reactDocGenTypescriptOptions?: RectDocgenTypescriptOptions,
) => {
  const {
    transformProps = (tables: any[]) => {
      const byName =
        componentName &&
        tables.find(table => table.displayName === componentName);
      return byName ? byName : tables[0];
    },
    propFilter = {
      skipPropsWithoutDoc: false,
    },
    componentNameResolver = computeComponentName,
    shouldExtractLiteralValuesFromEnum = true,
    savePropValueAsString,
  } = reactDocGenTypescriptOptions || {};

  const parserOptions: ParserOptions = {
    propFilter,
    //@ts-ignore
    componentNameResolver,
    shouldExtractLiteralValuesFromEnum,
    savePropValueAsString,
  };
  const parser = withDefaultConfig(parserOptions);
  try {
    let docgenInfo = parser.parse(fileName);
    if (Array.isArray(docgenInfo) && docgenInfo.length > 0) {
      return transformProps(docgenInfo);
    }
  } catch (e) {
    console.error(
      `\nreact-docgen-typescript unable to parse ${componentName}: ${fileName}`,
    );
    return undefined;
  }
};
