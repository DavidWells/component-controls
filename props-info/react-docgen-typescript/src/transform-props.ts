import {
  PropTypes,
  PropType,
  TypeInformation,
  TypeValue,
} from '@component-controls/specification';
import { Props, PropItem } from 'react-docgen-typescript';

const cleanQuotes = (txt?: string) =>
  typeof txt === 'string' ? txt.replace(/['"]+/g, '') : txt;

export const transformProps = (props: Props): PropTypes => {
  return Object.keys(props).reduce((acc, name) => {
    const rdProp: PropItem = props[name];
    const prop: Partial<PropType> = {};
    if (rdProp.description) {
      prop.description = rdProp.description;
    }
    if (rdProp.parent?.name) {
      prop.parentName = rdProp.parent?.name;
    }
    if (rdProp.defaultValue !== null && rdProp.defaultValue !== undefined) {
      prop.defaultValue = rdProp.defaultValue.value || rdProp.defaultValue;
    }
    const typeName = rdProp.type.name || '';
    let type: Partial<TypeInformation> = {};
    if (rdProp.required) {
      type.required = rdProp.required;
    }

    if (typeName === 'enum') {
      type.name = 'enum';
      type.value = rdProp.type.value.map(({ value }: any) => {
        const val = cleanQuotes(value);
        return {
          name: typeof val,
          value: val,
        };
      });
    } else if (typeName.endsWith('[]')) {
      type.name = 'array';
      type.value = [{ name: typeName.split('[]')[0] as TypeValue }];
    } else if (typeName.match(/\(([^)]+)\)/)) {
      type.name = 'function';
    } else if (typeName.match(/\{[^{}]*}/)) {
      type.name = 'object';
      type.raw = typeName;
    } else {
      const enumSplit = typeName.split(' | ');
      if (enumSplit.length > 1) {
        type.name = 'union';
        type.value = enumSplit.map(str => ({
          name: cleanQuotes(str),
        }));
      } else {
        type.name = typeName as TypeValue;
      }
    }
    const raw = rdProp.type.raw || typeName;
    if (type.name !== raw) {
      type.raw = raw;
    }
    prop.type = type as TypeInformation;
    return { ...acc, [name]: prop };
  }, {});
};
