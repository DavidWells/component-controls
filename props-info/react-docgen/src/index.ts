import { extractDocgenInfo, RectDocgenOptions } from './react-docgen';
import { transformProps } from './transform-props';

export default (
  fileName: string,
  source?: string,
  options?: RectDocgenOptions,
) => {
  const propTable = extractDocgenInfo(fileName, source, options);
  return {
    ...propTable,
    props: transformProps(propTable.props),
  };
};
