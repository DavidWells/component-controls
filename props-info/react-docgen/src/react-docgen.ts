import * as fs from 'fs';
import { parse } from 'react-docgen';
import { ComponentInfo } from '@component-controls/core';

import { RectDocgenOptions, RdPropInfoRecord } from './types';

export const extractDocgenInfo = (
  fileName: string,
  componentName?: string,
  source?: string,
  reactDocGenOptions?: RectDocgenOptions,
): ({ props: RdPropInfoRecord } & ComponentInfo) | undefined => {
  const { resolver, handlers, options } = reactDocGenOptions || {};
  const src = source ? source : fs.readFileSync(fileName, 'utf8');
  try {
    return parse(src, resolver, handlers, { filename: fileName, ...options });
  } catch (e) {
    console.error(
      `\nreact-docgen unable to parse ${componentName}: ${fileName}`,
    );
    return undefined;
  }
};
