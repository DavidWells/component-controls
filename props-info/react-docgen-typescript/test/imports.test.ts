import path from 'path';
import { run } from '../src/index';

export type LoadTestCallbackFn = (fileName: string) => any;

export const loadTestFiles = (fileName: string, componentName: string) => {
  const filePath = path.resolve(__dirname, fileName);
  console.log(filePath);
  it(fileName, async () => {
    expect(await run()(filePath, componentName)).toMatchSnapshot();
  });
};

describe('imports', () => {
  loadTestFiles('../../../ui/blocks/src/Story/Story.tsx', 'Story');
});
