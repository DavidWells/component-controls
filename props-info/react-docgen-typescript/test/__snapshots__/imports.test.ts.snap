// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`imports ../../../ui/blocks/src/Story/Story.tsx 1`] = `
Object {
  "description": "",
  "displayName": "Story",
  "methods": Array [],
  "props": Object {
    "collapsible": Object {
      "description": "if false, will nothave a collapsible frame.",
      "parentName": "BlockContainerProps",
      "type": Object {
        "name": "boolean",
        "raw": "boolean",
      },
    },
    "id": Object {
      "description": "id of the story",
      "parentName": "StoryInputProps",
      "type": Object {
        "name": "string",
        "raw": "string",
      },
    },
    "name": Object {
      "description": "alternatively you can use the name of a story to load from an external file",
      "parentName": "StoryInputProps",
      "type": Object {
        "name": "string",
        "raw": "string",
      },
    },
    "sxStyle": Object {
      "description": "theme-ui styling object for Block Box",
      "parentName": "BlockContainerProps",
      "type": Object {
        "name": "SystemStyleObject",
        "raw": "SystemStyleObject",
      },
    },
    "title": Object {
      "description": "optional section title for the block.",
      "parentName": "BlockContainerProps",
      "type": Object {
        "name": "string",
        "raw": "string",
      },
    },
  },
}
`;
